src/urg_node_driver.cpp

```diff
@@ -139,204 +155,308 @@ bool UrgNode::updateStatus()
+    ss << "\n  " << deleted_parameter.name;
+  }
+  ss << "\n";
+  RCLCPP_INFO(this->get_logger(), ss.str().c_str());
+
+  // Concat the new parameters (I guess there shouldn't be any though)
+  // with the changed parameters into one vector
+  std::vector<rcl_interfaces::msg::Parameter> parameter_vec = event->new_parameters;
+  parameter_vec.insert(parameter_vec.end(),
+    event->changed_parameters.begin(), event->changed_parameters.end());
+
+  // Some parameter change require to stop and start the driver to be applied
+  bool restart_required(false);
+
+  // Get each parameter one by one, the param_change_callback should leave us
+  // only with valid parameters
+  for (auto parameter : parameter_vec) {
+    if (parameter.name.compare("laser_frame_id") == 0) {
+      laser_frame_id_ = parameter.value.string_value;
+
+    } else if (parameter.name.compare("error_limit") == 0) {
+      error_limit_ = parameter.value.integer_value;
+
+    } else if (parameter.name.compare("default_user_latency") == 0) {
+      default_user_latency_ = parameter.value.integer_value + parameter.value.double_value;
+
+    } else if (parameter.name.compare("angle_min") == 0) {
+      angle_min_ = parameter.value.integer_value + parameter.value.double_value;
+      restart_required = true;
+
+    } else if (parameter.name.compare("angle_max") == 0) {
+      angle_max_ = parameter.value.integer_value + parameter.value.double_value;
+      restart_required = true;
+
+    } else if (parameter.name.compare("cluster") == 0) {
+      cluster_ = parameter.value.integer_value;
+      restart_required = true;
+
+    } else if (parameter.name.compare("skip") == 0) {
+      skip_ = parameter.value.integer_value;
+      restart_required = true;
+
+    } else {
+      RCLCPP_ERROR(
+        this->get_logger(), "The parameter %s is not part of the reconfigurable parameters.",
+        parameter.name.c_str());
+    }
+  }
+
+  // Apply the parameter changes
+  if (restart_required) {
+    std::unique_lock<std::mutex> lock(lidar_mutex_);
     urg_->stop();
-    ROS_INFO("Stopped data due to reconfigure.");
+    urg_->setAngleLimitsAndCluster(angle_min_, angle_max_, cluster_);
 
-    // Change values that required stopping
-    urg_->setAngleLimitsAndCluster(config.angle_min, config.angle_max, config.cluster);
-    urg_->setSkip(config.skip);
+    freq_min_ = 1.0 / (urg_->getScanPeriod() * (skip_ + 1));
+    urg_->setSkip(skip_);
 
-    try
-    {
+    try {
       urg_->start();
-      ROS_INFO("Streaming data after reconfigure.");
-    }
-    catch (std::runtime_error& e)
-    {
-      ROS_FATAL("%s", e.what());
-      ros::Duration(1.0).sleep();
-      ros::shutdown();
-      return false;
+      RCLCPP_INFO(this->get_logger(), "Streaming data after reconfigure.");
+    } catch (const std::runtime_error & e) {
+      RCLCPP_FATAL(this->get_logger(), "Error while reconfiguring : %s", e.what());
+      rclcpp::sleep_for(std::chrono::seconds(1));
+      rclcpp::shutdown();
```

Comment:
This feels like a layering violation.  Particularly in the case where this is a composable node in the future, we don't necessarily want to shutdown the whole world, we just want to quit ourselves.  So maybe it makes more sense to just re-raise the exception?

Reply:
done